File: part2/node/CloudantUserStore.js
'use strict';

const cloudant = require('cloudant');

/**
 * This class is used to store users and their state (Watson Conversation context) in Cloudant.
 */
class CloudantUserStore {

    /**
     * Creates a new instance of CloudantUserStore.
     * @param {string} cloudantUrl - The url of the of cloudant instance to connect to
     * @param {string} dbName - The name of the database to use
     */
    constructor(cloudantUrl, dbName) {
        this.cloudantUrl = cloudantUrl;
        this.cloudant = cloudant({
            url: cloudantUrl,
            plugin:'promises'
        });
        this.dbName = dbName;
        this.db = null;
    }

    /**
     * Creates and initializes the database.
     * @returns {Promise.<null|error>} - Null if fulfilled, or an error if rejected 
     */
    init() {
        console.log('Getting user database...');
        return this.cloudant.db.list()
            .then((dbNames) => {
                let exists = false;
                for (let dbName of dbNames) {
                    if (dbName == this.dbName) {
                        exists = true;
                    }
                }
                if (!exists) {
                    console.log(`Creating user database ${this.dbName}...`);
                    return this.cloudant.db.create(this.dbName);
                }
                else {
                    return Promise.resolve();
                }
            })
            .then(() => {
                this.db = this.cloudant.db.use(this.dbName);
                return Promise.resolve();
            });
    }

    /**
     * Adds a new user to Cloudant if a user with the specified ID does not already exist.
     * @param userId - The ID of the user (Slack ID, or unique ID associated with the WebSocket client)
     * @returns {Promise.<object|error>} - The user doc stored in Cloudant if fulfilled, or an error if rejected 
     */
    addUser(userId) {
        return this.db.get(userId)
            .then((userDoc) => {
                return Promise.resolve(userDoc);
            })
            .catch((err) => {
                if (err.statusCode == 404) {
                    let doc = { _id: userId };
                    return this.db.insert(doc)
                        .then((body) => {
                            doc._id = body.id;
                            doc._rev = body.rev;
                            return Promise.resolve(doc);
                        });
                }
                else {
                    return Promise.reject(err);
                }
            });
    }

    /**
     * Adds a new user to Cloudant if a user with the specified ID does not already exist.
     * @param userId - The ID of the user (Slack ID, or unique ID associated with the WebSocket client)
     * @returns {Promise.<TResult>} - The udpated user doc stored in Cloudant if fulfilled, or an error if rejected 
     */
    updateUser(user, context) {
        return this.db.get(user._id)
            .then((userDoc) => {
                let doc = userDoc;
                doc.conversationContext = context;
                return this.db.insert(doc)
                    .then((body) => {
                        doc._id = body.id;
                        doc._rev = body.rev;
                        return Promise.resolve(doc);
                    });
            });
    }
}

module.exports = CloudantUserStore;

File: part2/python/cloudant_user_store.py
import time

from cloudant.client import Cloudant
from cloudant.query import Query


class CloudantUserStore(object):

    def __init__(self, cloudant_username, cloudant_password, cloudant_url, db_name):
        """
        Creates a new instance of CloudantUserStore.
        Parameters
        ----------
        cloudant_username - The username for the cloudant instance
        cloudant_password - The password for the cloudant instance
        cloudant_url - The url of the of cloudant instance to connect to
        db_name - The name of the database to use
        """
        if cloudant_url.find('@') > 0:
            prefix = cloudant_url[0:cloudant_url.find('://')+3]
            suffix = cloudant_url[cloudant_url.find('@')+1:]
            cloudant_url = '{}{}'.format(prefix, suffix)
        self.client = Cloudant(cloudant_username, cloudant_password, url=cloudant_url)
        self.db_name = db_name

    def init(self):
        """
        Creates and initializes the database.
        """
        try:
            self.client.connect()
            print('Getting user database...')
            if self.db_name not in self.client.all_dbs():
                print('Creating user database {}...'.format(self.db_name))
                self.client.create_database(self.db_name)
            else:
                print('User database {} exists.'.format(self.db_name))
        finally:
            self.client.disconnect()

    # User

    def add_user(self, user_id):
        """
        Adds a new user to Cloudant if a user with the specified ID does not already exist.
        Parameters
        ----------
        user_id - The ID of the user (typically the ID returned from Slack)
        """
        try:
            self.client.connect()
            return self.client[self.db_name][user_id]
        except KeyError:
            db = self.client[self.db_name]
            doc = {
                '_id': user_id,
                'conversation_context': {}
            }
            return db.create_document(doc)
        finally:
            self.client.disconnect()

    def update_user(self, user, context):
        """
        Updates the user in Cloudant with the latest Watson Conversation context.
        Parameters
        ----------
        userId - The user doc stored in Cloudant
        context - The Watson Conversation context
        """
        try:
            self.client.connect()
            db = self.client[self.db_name]
            doc = db[user['_id']]
            doc['conversation_context'] = context
            return doc.save()
        finally:
            self.client.disconnect()